packages/@aws-cdk/aws-s3/lib/notifications-resource/lambda-source/src/index.py

```diff
@@ -0,0 +1,72 @@
+import boto3  # type: ignore
+import copy
+import json
+import urllib.request
+from typing import List
+
+s3 = boto3.client("s3")
+CONFIG_TYPES = ["TopicConfigurations", "QueueConfigurations", "LambdaFunctionConfigurations"]
+
+
+def handler(event: dict, context):
+    response_status = "SUCCESS"
+    error_message = ""
+    try:
+        props = event["ResourceProperties"]
+        bucket = props["BucketName"]
+        config = prepare_config(
+            current_config=s3.get_bucket_notification_configuration(Bucket=bucket),
+            in_config=props["NotificationConfiguration"],
+            old_config=event.get("OldResourceProperties", {}).get("NotificationConfiguration", {}),
+            merge=event["RequestType"] != "Delete",
+        )
+        s3.put_bucket_notification_configuration(Bucket=bucket, NotificationConfiguration=config)
+    except Exception as e:
+        print("Failed to put bucket notification configuration:", e)
+        response_status = "FAILED"
+        error_message = f"Error: {str(e)}. "
+    finally:
+        submit_response(event, context, response_status, error_message)
+
+
+def prepare_config(current_config: dict, in_config: dict, old_config: dict, merge: bool) -> dict:
+    config = copy.deepcopy(current_config)
+    config.pop("ResponseMetadata", None)
+    for config_type in CONFIG_TYPES:
+        config.setdefault(config_type, [])
+        in_config.setdefault(config_type, [])
+        filter_config(config, config_type, in_config, old_config)
```

Comment:
I'd much rather this look something like:

```suggestion
        config[config_type] = find_unmanaged_notifications(current_config, config_type, in_config, old_config)
```

Similarly to how you changed `prepare_config` to return instead of mutating.

Reply:
@iliapolo i am not sure what you are aiming for? This is functional and fully tested. Maybe you can put up the suggested changes so i can understand this better.

Reply:
I'm aiming for not having function mutate their own arguments, I find this type of flow confusing and bug prone. 

To achieve it we just need to make `filter_config` return the array, and have the caller assign it to `config` (which is not an argument, it's created inside the function and returned).

Reply:
@michaelbrewer Its not a deal breaker, if you don't have capacity i'm ok with skipping it, your call :)

Basically only the additions to `IBucket` interfaces are needed before we can merge.

Reply:
@iliapolo  I have make the `IBucket` changes. Modifying the lambda is time consuming as I have to go all of the integration json files and manually update them and if I get it wrong, I have to wait 30mins for the build to fail.